From faa60e0be933f49ac66eda2703a42dabf945d547 Mon Sep 17 00:00:00 2001
From: Artem Savkov <asavkov@redhat.com>
Date: Tue, 2 Dec 2014 13:02:05 -0500
Subject: [kernel] printk/register_console: prevent adding the same console twice

Message-id: <1417525325-25349-1-git-send-email-asavkov@redhat.com>
Patchwork-id: 101028
O-Subject: [RHEL7.1 PATCH BZ 1169766] register_console: prevent adding the same console twice
Bugzilla: 1169766
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1169766
Upstream Status: commited.
Build Info: https://brewweb.devel.redhat.com/taskinfo?taskID=8314482
Tested: Manually by myself

This is a backport of this commit from linux.git:

  commit 16cf48a6d3e8f9ebe3c3231c12cbe4b0c4ed4d24
  Author: Andreas Biessmann <andreas@biessmann.de>
  Date:   Fri Aug 2 12:23:34 2013 +0200

      register_console: prevent adding the same console twice

      This patch guards the console_drivers list to be corrupted. The
      for_each_console() macro insist on a strictly forward list ended by NULL:

       con0->next->con1->next->NULL

      Without this patch it may happen easily to destroy this list for example by
      adding 'earlyprintk' twice, especially on embedded devices where the early
      console is often a single static instance.  This will result in the following
      list:

       con0->next->con0

      This in turn will result in an endless loop in console_unlock() later on by
      printing the first __log_buf line endlessly.

      Signed-off-by: Andreas Biessmann <andreas@biessmann.de>
      Cc: Kay Sievers <kay@vrfy.org>
      Cc: Ben Hutchings <ben@decadent.org.uk>
      Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Artem Savkov <asavkov@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/printk.c b/kernel/printk.c
index 81ef6e3..6881a4f 100644
--- a/kernel/printk.c
+++ b/kernel/printk.c
@@ -2254,6 +2254,13 @@ void register_console(struct console *newcon)
  unsigned long flags;
  struct console *bcon = NULL;
 
+ if (console_drivers)
+  for_each_console(bcon)
+   if (WARN(bcon == newcon,
+     "console '%s%d' already registered\n",
+     bcon->name, bcon->index))
+    return;
+
  /*
   * before we register a new CON_BOOT console, make sure we don't
   * already have a valid console
-- 
1.7.1